upf/upf_ipfix.c

```diff
@@ -0,0 +1,693 @@
+  vlib_main_t *vm = fm->vlib_main;
+
+  if (fm->disabled || f->ipfix_context_index == (u32) ~0)
+    return;
+
+  /* TODO: unidirectional export */
+  upf_ipfix_export_entry (vm, f, FT_ORIGIN, now);
+  upf_ipfix_export_entry (vm, f, FT_REVERSE, now);
+
+  upf_unref_ipfix_context_by_index (f->ipfix_context_index);
+}
+
+static inline void
+upf_ipfix_context_mk_key(bool is_ip4,
+			 upf_ipfix_policy_t policy,
+			 const ip_address_t * collector_ip,
+			 clib_bihash_kv_24_8_t * kv)
+{
+  kv->key[2] = (policy << 2) | (is_ip4 ? 2 : 0);
+  /* Avoid any ambiguity due to IPv4/IPv6 zero IPs */
+  if (!collector_ip || ip_address_is_zero (collector_ip))
+    {
+      kv->key[0] = 0;
+      kv->key[1] = 0;
+    }
+  else
+    {
+      kv->key[0] = collector_ip->ip.as_u64[0];
+      kv->key[1] = collector_ip->ip.as_u64[1];
+      if (collector_ip->version == AF_IP4)
+	kv->key[2] |= 1;
+    }
+}
+
+u32
+upf_ref_ipfix_context (bool is_ip4,
+		       upf_ipfix_policy_t policy,
+		       const ip_address_t * collector_ip)
+{
+  int rv;
+  vlib_thread_main_t *tm = &vlib_thread_main;
+  upf_ipfix_main_t *fm = &upf_ipfix_main;
+  clib_bihash_kv_24_8_t kv, value;
+  upf_ipfix_protocol_context_t *context;
+  /* Decide how many worker threads we have */
+  u32 num_threads = 1 /* main thread */  + tm->n_threads;
+
+  upf_ipfix_context_mk_key(is_ip4, policy, collector_ip, &kv);
+
+  if (PREDICT_TRUE
+      (!clib_bihash_search_24_8
+       (&fm->context_by_key, &kv, &value)))
+    {
+      context = pool_elt_at_index (fm->contexts, value.value);
+      context->ref_count++;
+      return value.value;
+    }
+
+  pool_get_zero (fm->contexts, context);
+
+  vec_validate (context->buffers_per_worker, num_threads - 1);
+  vec_validate (context->frames_per_worker, num_threads - 1);
+  vec_validate (context->next_record_offset_per_worker,
+		num_threads - 1);
+  kv.value = context - fm->contexts;
+  clib_bihash_add_del_24_8 (&fm->context_by_key, &kv, 1);
+
+  context->policy = policy;
+  context->is_ip4 = is_ip4;
+  /* lookup the exporter a bit later */
+  context->exporter_index = (u32) ~0;
+  context->ref_count = 1;
+  if (collector_ip)
+    ip_address_copy (&context->collector_ip, collector_ip);
+
+  rv = upf_ipfix_report_add_del (fm, UPF_IPFIX_OBSERVATION_DOMAIN_ID,
+				 kv.value, &context->template_id,
+				 is_ip4, true);
+  if (rv)
+      clib_warning ("couldn't add IPFIX report, perhaps "
+		    "the exporter has been deleted?");
+
+  return context - fm->contexts;
+}
+
+void
+upf_ref_ipfix_context_by_index (u32 cidx)
+{
+  upf_ipfix_main_t *fm = &upf_ipfix_main;
+  upf_ipfix_protocol_context_t *context;
+
+  context = pool_elt_at_index (fm->contexts, cidx);
+  context->ref_count++;
+}
+
+void
+upf_unref_ipfix_context_by_index (u32 cidx)
+{
+  int rv;
+  upf_ipfix_main_t *fm = &upf_ipfix_main;
+  clib_bihash_kv_24_8_t kv;
+  upf_ipfix_protocol_context_t *context;
+
+  context = pool_elt_at_index (fm->contexts, cidx);
+  if (--context->ref_count)
```

Comment:
upf_unref_ipfix_context_by_index can could be invoked by multiple threads for the same flow. So updating the ref_count should use a atomic function (probably in other places as well).

Reply:
Switched to atomic func